app/uk/gov/hmrc/cgtpropertydisposals/models/bpr/BusinessPartnerRecordRequest.scala

```diff
@@ -40,12 +43,18 @@ object BusinessPartnerRecordRequest {
 
     def fold[A](
       ifTrust: TrustBusinessPartnerRecordRequest => A,
-      ifIndividual: IndividualBusinessPartnerRecordRequest => A): A = r match {
+      ifIndividual: IndividualBusinessPartnerRecordRequest => A
+    ): A = r match {
       case t: TrustBusinessPartnerRecordRequest      => ifTrust(t)
       case i: IndividualBusinessPartnerRecordRequest => ifIndividual(i)
     }
 
-    def id: Either[SAUTR, NINO] = fold(t => Left(t.id), _.id)
+    def foldOnId[A](
+      ifTrn: TRN => A,
+      ifSautr: SAUTR => A,
+      ifNino: NINO => A
+    ): A =
+      fold(_.id.fold(ifTrn, ifSautr), _.id.fold(ifSautr, ifNino))
```

Comment:
Again, this is obfuscated code. I understand why you have done it but this can be implemented using case analysis. Granted that would be more lines of code but it would make it much easier for others to read and understand.